crates/lib3h/src/rrdht_util/arc.rs

```diff
@@ -0,0 +1,336 @@
+use crate::rrdht_util::Location;
+
+pub const ARC_LENGTH_MAX: u64 = 0x100000000;
+pub const ARC_RADIUS_MAX: u32 = 0x80000001;
+
+/// An rrdht "arc" indicates a range on the u32 "location" circle
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct Arc {
+    start: Location,
+    // length must be a u64 because we need to be able to represent
+    // both a zero length arc with length 0,
+    // and also a full arc with length one beyond the u32 max
+    length: u64,
+}
+
+impl std::fmt::Display for Arc {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "32r{:08x}:{:09x}", u32::from(self.start), self.length)
+    }
+}
+
+impl std::fmt::Debug for Arc {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("Arc").field(&format!("{}", self)).finish()
+    }
+}
+
+impl Arc {
+    /// construct a new Arc instance given a start Location and a length
+    /// if length > ARC_LENGTH_MAX it will be clipped to that value
+    pub fn new(start: Location, length: u64) -> Self {
+        let length = std::cmp::min(length, ARC_LENGTH_MAX);
+        Self { start, length }
+    }
+
+    /// construct a new Arc instance given a center Location and a radius
+    /// radius as in half the arc length
+    pub fn new_radius(center: Location, radius: u32) -> Self {
+        if radius == 0 {
+            return Arc::new(center, 0);
+        }
+
+        let radius = std::cmp::min(radius, ARC_RADIUS_MAX);
+
+        // we need to subtract 1 from the radius in the following
+        // calculations to ensure that:
+        // - 0 means no coverage
+        // - 1 means coverage only including the center point
+        // - any thing obove expand out to the sides
```

Comment:
This comment looks like it could make sense to promote it to a rustdoc comment on the whole type..

Reply:
:+1: - done